### PR TITLE
Web worker for amp-bind

### LINKED ISSUE #6910: amp-bind: Improve init performance
Related to #6199.

Several slow tasks are necessary during initialization. We should speed it up where possible and chunk/amortize across multiple frames otherwise:

- [x] DOM scan for bindings
- [ ] Expression parsing and AST generation

### PR DESCRIPTION
**Pull request by choumx**
Fixes #6910, partial for #6199.

- Implements a `Worker` in a new entry point compiled to `dist/ww.js`
- Adds `AmpWorker` class that wraps the `Worker` with a Promise-based API
- Use in `amp-bind` to parse and evaluate expressions on background thread
- Flagged under new "web-worker" experiment

### CODE REVIEW

**Comment by choumx** [comment]
/to @cramforce @jridgewell /cc @kmh287 

PTAL, working on tests now.

File: extensions/amp-bind/0.1/bind-impl.js [hunk 1]
```diff
@@ -142,16 +156,27 @@ export class Bind {
...
       this.boundElements_ = boundElements;
-
       Object.assign(this.expressionToElements_, expressionToElements);
+      dev().fine(TAG, `Scanned ${bindings.length} bindings from ` +
+          `${boundElements.length} elements.`);
 
-      this.evaluator_ = new BindEvaluator();
-      const parseErrors = this.evaluator_.setBindings(bindings);
+      // Parse on web worker if experiment is enabled.
+      if (this.workerExperimentEnabled_) {
```
Comment (cramforce) [inline extensions/amp-bind/0.1/bind-impl.js]: Are you making this an experiment to not disrupt your other development, for easy benchmarking, or is there a different reason?

File: extensions/amp-bind/0.1/bind-impl.js [hunk 2]
```diff
@@ -345,18 +377,27 @@ export class Bind {
...
         let width, height;
 
         boundProperties.forEach(boundProperty => {
-          const expressionString = boundProperty.expressionString;
-          const newValue = results[expressionString];
+          const {property, expressionString, previousResult} =
+              boundProperty;
+
+          // Rewrite attribute value if necessary.
+          // This is not done in the worker since it relies on `url#parseUrl`,
```
Comment (cramforce) [inline extensions/amp-bind/0.1/bind-impl.js]: Definitely fine for now. The `URL` API may be exposed to workers.

File: src/web-worker/amp-worker.js [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+  /**
+   * Sends a method invocation request to the worker and returns a Promise.
+   * @param {string} method
+   * @param {!Array} args
+   * @return {!Promise}
+   * @private
+   */
+  sendMessage_(method, args) {
+    const promise = new Promise((resolve, reject) => {
+      if (!this.messages_[method]) {
```
Comment (cramforce) [inline src/web-worker/amp-worker.js]: This works, but it might be simpler to have a map of outstanding messages with a simple running index that you increment for each message.

File: src/web-worker/amp-worker.js [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+
+    // Clean up array if there are no more messages in flight for this method.
+    let empty = true;
+    for (let i = 0; i < invocations.length && empty; i++) {
+      if (invocations[i] !== undefined) {
+        empty = false;
+      }
+    }
+    if (empty) {
+      delete this.messages_[method];
```
Comment (cramforce) [inline src/web-worker/amp-worker.js]: As explained above, this is simpler with a running index and one map independent of method name.

File: src/web-worker/web-worker.js [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+/** @private {BindEvaluator} */
+let evaluator_;
+
+self.addEventListener('message', function(event) {
+  const {method, args, id} = /** @type {ToWorkerMessageDef} */ (event.data);
+
+  let returnValue;
+
+  switch (method) {
```
Comment (cramforce) [inline src/web-worker/web-worker.js]: Please add TODO and file issue to add error reporting to this.

We can slightly generify the code in `src/service-worker/error-reporting.js`

File: src/web-worker/web-worker.js [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+      returnValue = evaluator_.setBindings.apply(evaluator_, args);
+      break;
+
+    case 'bind.evaluate':
+      if (evaluator_) {
+        returnValue = evaluator_.evaluate.apply(evaluator_, args);
+      } else {
+        throw new Error(`${method}: BindEvaluator is not initialized.`);
+      }
+      break;
```
Comment (cramforce) [inline src/web-worker/web-worker.js]: Throw on default case.

File: extensions/amp-bind/0.1/bind-impl.js (hunk 2 above)
Comment (choumx) [inline extensions/amp-bind/0.1/bind-impl.js]: The problem is that [`parseUrl` uses `document.createElement`](https://github.com/ampproject/amphtml/blob/master/src/url.js#L54), which IIUC is unavailable to workers.

To workaround, we could either change this implementation to not use DOM or even use `Worker#postMessage` to delegate this to main page. Thoughts?

File: src/web-worker/amp-worker.js [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+ * @param {string} method
+ * @param {!Array=} opt_args
+ * @return {!Promise}
+ */
+export function invokeWebWorker(win, method, opt_args) {
+  if (!isExperimentOn(win, TAG)) {
+    return Promise.reject(`Experiment "${TAG}" is disabled.`);
+  }
+  if (!win.Worker) {
+    return Promise.reject('Worker not supported in window: ' + win);
```
Comment (jridgewell) [inline src/web-worker/amp-worker.js]: The `+ win` will generate a useless `"[object Window]"` string.

File: gulpfile.js [hunk 8]
```diff
@@ -888,6 +891,24 @@ function buildSw(options) {
...
+ *
+ * @param {!Object} options
+ */
+function buildWebWorker(options) {
+  $$.util.log('Bundling ww.js');
+  var opts = Object.assign({}, options);
+
+  return compileJs('./src/web-worker/', 'web-worker.js', './dist/', {
+    toName: 'ww.max.js',
+    minifiedName: 'ww.js',
```
Comment (jridgewell) [inline gulpfile.js]: This will require special handling on our servers. /cc @vitaliybl

File: src/web-worker/amp-worker.js [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+
+  /**
+   * Sends a method invocation request to the worker and returns a Promise.
+   * @param {string} method
+   * @param {!Array} args
+   * @return {!Promise}
+   * @private
+   */
+  sendMessage_(method, args) {
+    const promise = new Promise((resolve, reject) => {
```
Comment (jridgewell) [inline src/web-worker/amp-worker.js]: Nit: return here.

File: src/web-worker/amp-worker.js [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+   * Receives the result of a method invocation from the worker and resolves
+   * the Promise returned from the corresponding `sendMessage_()` call.
+   * @param {!MessageEvent} event
+   * @private
+   */
+  receiveMessage_(event) {
+    const {method, returnValue, id} =
+        /** @type {FromWorkerMessageDef} */ (event.data);
+
+    const message = this.messages_[id];
```
Comment (jridgewell) [inline src/web-worker/amp-worker.js]: I hate this `id` based on `messages.length`. Let's use a monotonic `AmpWorker#id`, increasing on every message. Then do a binary search when we receive a message. (We're already doing `O(n)` on every message receipt)

File: src/web-worker/amp-worker.js [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+        `(${method}, ${id}), expected ${message.method}.`);
+
+    message.resolve(returnValue);
+    this.messages_[id] = undefined;
+
+    // Clean up array if there are no more pending messages.
+    let empty = true;
+    for (let i = 0; i < this.messages_.length && empty; i++) {
+      if (this.messages_[i] !== undefined) {
+        empty = false;
```
Comment (jridgewell) [inline src/web-worker/amp-worker.js]: Break afterwards.

File: src/web-worker/amp-worker.js [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+      }
+    }
+    if (empty) {
+      this.messages_ = [];
```
Comment (jridgewell) [inline src/web-worker/amp-worker.js]: Nit: `message.length = 0` will  truncate and save the array instance.

File: src/web-worker/web-worker.js [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+let evaluator_;
+
+self.addEventListener('message', function(event) {
+  const {method, args, id} = /** @type {ToWorkerMessageDef} */ (event.data);
+
+  let returnValue;
+
+  // TODO(choumx): Add error reporting.
+  switch (method) {
+    case 'bind.initialize':
```
Comment (jridgewell) [inline src/web-worker/web-worker.js]: Nit: I dislike massive switch statements. Can we split this these into functions (we can keep the switch, then), or split them into functions and use some sort of map from "method" -> function.

File: src/web-worker/amp-worker.js (hunk 10 above)
Comment (choumx) [inline src/web-worker/amp-worker.js]: Splice completed messages from the array and binary search to find them? Seems more complex and slower, I don't see the benefit.

> ...hate leads to suffering, suffering leads to the dark side.

😄

File: src/web-worker/amp-worker.js (hunk 11 above)
Comment (choumx) [inline src/web-worker/amp-worker.js]: The termination condition contains `&& empty` and @camelburrito hates `break`.

File: src/web-worker/amp-worker.js (hunk 12 above)
Comment (choumx) [inline src/web-worker/amp-worker.js]: 👍

File: src/web-worker/web-worker.js (hunk 13 above)
Comment (choumx) [inline src/web-worker/web-worker.js]: Point taken. With only two cases at the moment, using a `<string, Function>` map is less readable IMO. Besides, this file will need to be significantly refactored to support other use cases in the future anyways.

File: gulpfile.js (hunk 8 above)
Comment (choumx) [inline gulpfile.js]: Good to know, I'll also reach out.

File: src/web-worker/amp-worker.js (hunk 11 above)
Comment (jridgewell) [inline src/web-worker/amp-worker.js]: Should be a while loop, then. 😛

File: src/web-worker/amp-worker.js (hunk 10 above)
Comment (jridgewell) [inline src/web-worker/amp-worker.js]: We have a never ending array this way...

File: src/web-worker/web-worker.js (hunk 13 above)
Comment (jridgewell) [inline src/web-worker/web-worker.js]: I mean something like:

```js
switch (method) {
  case "whatever":
    returnValue = whatever(...);
    break;
  case "other":
    returnValue = other(...);
    break;
}

File: src/web-worker/web-worker.js [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+      evaluator_ = new BindEvaluator();
+      returnValue = evaluator_.setBindings.apply(evaluator_, args);
+      break;
+
+    case 'bind.evaluate':
+      if (evaluator_) {
```
Comment (jridgewell) [inline src/web-worker/web-worker.js]: Do web workers terminate? If so, `evaluator_` could be undefined even if it was initialized previously.

Comment (cramforce) [inline src/web-worker/web-worker.js]: I don't think they do. But we'll find out via error reporting.

File: gulpfile.js (hunk 8 above)
Comment (cramforce) [inline gulpfile.js]: Also CC @erwinmombay: A new JS binary on the way.

File: src/web-worker/web-worker.js (hunk 14 above)
Comment (choumx) [inline src/web-worker/web-worker.js]: As far as I can tell, they only [terminate on request](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Using_web_workers#Terminating_a_worker).

File: src/web-worker/amp-worker.js (hunk 10 above)
Comment (choumx) [inline src/web-worker/amp-worker.js]: Only while messages are in flight which shouldn't be an issue for current functionality.

File: src/web-worker/amp-worker.js [hunk 15]
```diff
@@ -0,0 +1,144 @@
...
+        calculateEntryPointScriptUrl(location, 'ww', getMode().localDev);
+    /** @const @private {!Worker} */
+    this.worker_ = new win.Worker(url);
+    this.worker_.onmessage = this.receiveMessage_.bind(this);
+
+    /**
+     * Array of in-flight messages pending response from worker.
+     * @const @private {!Array<(PendingMessageDef|undefined)>}
+     */
+    this.messages_ = [];
```
Comment (cramforce) [inline src/web-worker/amp-worker.js]: Sorry, missed this earlier: This should be an object. Makes the deletion after messages are done so much easier.

Comment (choumx) [inline src/web-worker/amp-worker.js]: Good idea, done.

File: src/web-worker/amp-worker.js (hunk 10 above)
Comment (choumx) [inline src/web-worker/amp-worker.js]: Changed this to use a counter per Malte's comment above.

File: src/web-worker/amp-worker.js [hunk 16]
```diff
@@ -0,0 +1,141 @@
...
+  /**
+   * Sends a method invocation request to the worker and returns a Promise.
+   * @param {string} method
+   * @param {!Array} args
+   * @return {!Promise}
+   * @private
+   */
+  sendMessage_(method, args) {
+    return new Promise((resolve, reject) => {
+      this.messages_[this.counter_] = {method, resolve, reject};
```
Comment (cramforce) [inline src/web-worker/amp-worker.js]: While we don't need to be thread safe, this is still dangerous if anything throws an exception.

Something like `const messageIndex = this.counter_++` would be more obvious, I think.

Comment (choumx) [inline src/web-worker/amp-worker.js]: Done.

File: gulpfile.js (hunk 8 above)
Comment (erwinmombay) [inline gulpfile.js]: got it.

File: extensions/amp-bind/0.1/bind-impl.js [hunk 17]
```diff
@@ -99,11 +105,20 @@ export class Bind {
...
+    this.initialized_ = false;
+
+    /**
      * True if a digest is triggered before scan for bindings completes.
      * @private {boolean}
      */
     this.digestQueuedAfterScan_ = false;
 
+    /** @const @private {boolean} */
+    this.workerExperimentEnabled_ = isExperimentOn(this.win_, 'web-worker');
```
Comment (jridgewell) [inline extensions/amp-bind/0.1/bind-impl.js]: We also need to check that web workers are supported. Might be better to export a function from the `amp-worker.js` module.

File: src/web-worker/amp-worker.js [hunk 18]
```diff
@@ -0,0 +1,140 @@
...
+
+    const url =
+        calculateEntryPointScriptUrl(location, 'ww', getMode().localDev);
+    /** @const @private {!Worker} */
+    this.worker_ = new win.Worker(url);
+    this.worker_.onmessage = this.receiveMessage_.bind(this);
+
+    /**
+     * Array of in-flight messages pending response from worker.
+     * @const @private {!Object<number, (PendingMessageDef|undefined)>}
```
Comment (jridgewell) [inline src/web-worker/amp-worker.js]: No need for the `|undefined`.

File: src/web-worker/amp-worker.js [hunk 19]
```diff
@@ -0,0 +1,140 @@
...
+   * Sends a method invocation request to the worker and returns a Promise.
+   * @param {string} method
+   * @param {!Array} args
+   * @return {!Promise}
+   * @private
+   */
+  sendMessage_(method, args) {
+    return new Promise((resolve, reject) => {
+      const id = this.counter_++;
+      this.messages_[id] = {method, resolve, reject};
```
Comment (jridgewell) [inline src/web-worker/amp-worker.js]: `reject` isn't used?

File: src/web-worker/web-worker.js [hunk 20]
```diff
@@ -0,0 +1,58 @@
...
+/** @private {BindEvaluator} */
+let evaluator_;
+
+self.addEventListener('message', function(event) {
+  const {method, args, id} = /** @type {ToWorkerMessageDef} */ (event.data);
+
+  let returnValue;
+
+  // TODO(choumx): Add error reporting.
+  switch (method) {
```
Comment (jridgewell) [inline src/web-worker/web-worker.js]: Let's wrap this in a try-catch, then post back an error if something goes wrong.

File: src/web-worker/amp-worker.js (hunk 19 above)
Comment (choumx) [inline src/web-worker/amp-worker.js]: May be useful for returning warnings or errors to the caller. Error reporting for the worker is next: #7453

File: src/web-worker/web-worker.js (hunk 20 above)
Comment (choumx) [inline src/web-worker/web-worker.js]: Good idea. Tracked in #7453.

File: extensions/amp-bind/0.1/bind-impl.js (hunk 17 above)
Comment (choumx) [inline extensions/amp-bind/0.1/bind-impl.js]: Looks like [web workers are available](http://caniuse.com/#feat=webworkers) for all browsers we care about, so I don't think we should bother supporting the fallback case. It's still useful to check this in `amp-worker.js` so we can reject `invokeWebWorker` with a meaningful error message.

File: test/functional/web-worker/test-amp-worker.js [hunk 21]
```diff
@@ -39,15 +40,16 @@ describe('invokeWebWorker', () => {
     const fakeWorkerClass = () => fakeWorker;
     fakeWin = {Worker: fakeWorkerClass};
 
-    toggleExperiment(fakeWin, 'web-worker', true);
+    isExperimentOnStub = sandbox.stub(experiments, 'isExperimentOn');
```
Comment (jridgewell) [inline test/functional/web-worker/test-amp-worker.js]: This is supppppperrrrr icky, and not valid ES6. `import * as experiments` means `experiments` is not an mutable object.

Why was this changed?

Comment (choumx) [inline test/functional/web-worker/test-amp-worker.js]: Test was failing due to a recent change in the implementation of `toggleExperiment` and figured stubbing is less brittle. Is there no good way to stub exported methods?

Reverted and changed to avoid the source of the original error.

Comment (jridgewell) [inline test/functional/web-worker/test-amp-worker.js]: You have to export a setter method from the module:

```js
export function method() {}

export function setMethodForTesting(fn) {
  method = fn;
}

Comment (choumx) [inline test/functional/web-worker/test-amp-worker.js]: Ew. 😄 

Thanks for the review.
